management-interface.go

```diff
@@ -289,6 +293,301 @@ func NetworkHandler(w http.ResponseWriter, r *http.Request) {
+			exists = true
+		}
+	}
+	return exists
+}
+
+//addWpaNetwork - adds a new wpa network with specified ssid and password (if it doesn't already exist)
+func addWpaNetwork(ssid string, password string) error {
+	err := error(nil)
+	if ssid == "" {
+		err = errors.New("SSID must have a value")
+		log.Printf("Ssid must have a value")
+		return err
+	}
+
+	if doesWpaNetworkExist(ssid) {
+		err = errors.New("SSID already exists " + ssid)
+		log.Printf("Ssid %s already exits", ssid)
+		return err
+	}
+
+	var networkId int
+	networkId, err = addNewNetwork()
+	if err != nil {
+		return err
+	}
+
+	err = setWpaNetworkDetails(ssid, password, networkId)
+	if err != nil {
+		return err
+	}
+
+	err = saveWpaConfig()
+	reloadErr := reloadWpaConfig()
+	if err == nil { //probably wont happen
+		err = reloadErr
+	}
+	return err
+}
+
+//addNewNetwork - a a new network and returns the network id
+func addNewNetwork() (int, error) {
+	out, err := exec.Command("wpa_cli", "add_network").Output()
+	var networkId int = -1
+
+	if err != nil {
+		log.Printf("error executing wpa_cli add_network - error %s output %s", err.Error(), out)
+		return networkId, err
+	}
+	stdOut := string(out)
+
+	//get the networkid of the new networks from stdOut
+	scanner := bufio.NewScanner(strings.NewReader(stdOut))
+	scanner.Scan() //skip interface line
+	if scanner.Scan() {
+		line := scanner.Text()
+		networkId, err = strconv.Atoi(line)
+		if err != nil {
+			log.Printf("could not find network id - error %s from stdout %s", err.Error(), stdOut)
+			return -1, err
+		}
+	}
+	return networkId, err
+}
+
+//setWpaNetworkDetails - sets the ssid and password of the specified networkID
+func setWpaNetworkDetails(ssid string, password string, networkId int) error {
+	cmd := exec.Command("wpa_cli")
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		log.Printf("error getting stdin pipe from cnd -error %s", err.Error())
+		return err
+	}
+
+	defer stdin.Close()
+	io.WriteString(stdin, fmt.Sprintf("set_network %d ssid \"%s\"\n", networkId, ssid))
+	io.WriteString(stdin, fmt.Sprintf("set_network %d psk \"%s\"\n", networkId, password))
+	io.WriteString(stdin, fmt.Sprintf("enable_network %d\n", networkId))
+	io.WriteString(stdin, "quit\n")
+	var out []byte
+
+	out, err = cmd.CombinedOutput()
+
+	if err != nil {
+		log.Printf("error adding wpa network -error %s", err.Error())
+		return err
+	}
+
+	errOccured := hasErrorOccured(string(out))
+	if errOccured {
+		reloadWpaConfig()
+		err = errors.New("error setting new network")
+		log.Printf(err.Error())
+	}
+	return err
+}
+
+//reloadWpaConfig - executes wpa_cli reconfigure
+func reloadWpaConfig() error {
+	out, err := exec.Command("wpa_cli", "reconfigure").Output()
+	if err != nil {
+		log.Printf("error reloading config - error %s output %s", err.Error(), out)
+		return err
+	}
+
+	errOccured := hasErrorOccured(string(out))
+	if errOccured {
+		err = errors.New("error reloading config")
+		log.Printf(err.Error())
```

Comment:
and all the other places

Reply:
In this case is it not best to create the error so I can return it, for parents methods to acknowledge and error has occurred?